File: 4.8-bc/ts/stage-3-hexagonal/src/notifications/infra/postgres/notificationRepo.ts
import type { DbPool } from "../../../shared/db.js";
import type { Notification } from "../../domain/notification.js";
import type { Repository } from "../../domain/repository.js";

export class PgNotificationRepo implements Repository {
  constructor(private readonly db: DbPool) {}

  async save(n: Notification): Promise<void> {
    await this.db.query(
      `INSERT INTO notifications_messages (id, user_id, channel, payload, sent_at)
       VALUES ($1, $2, $3, $4, $5)`,
      [n.id, n.userId, n.channel, n.payload, n.sentAt],
    );
  }
}

File: 4.8-bc/py/stage-3-hexagonal/app/notifications/infra/postgres/notification_repo.py
"""Postgres adapter for Notifications Repository."""
from __future__ import annotations

from sqlalchemy import text
from sqlalchemy.ext.asyncio import async_sessionmaker

from app.notifications.domain.notification import Notification


class NotificationRepo:
    def __init__(self, session_maker: async_sessionmaker) -> None:
        self._session_maker = session_maker

    async def save(self, n: Notification) -> None:
        async with self._session_maker() as session:
            await session.execute(
                text(
                    "INSERT INTO notifications_messages "
                    "(id, user_id, channel, payload, sent_at) "
                    "VALUES (:id, :user_id, :channel, :payload, :sent_at)"
                ),
                {
                    "id": n.id,
                    "user_id": n.user_id,
                    "channel": str(n.channel),
                    "payload": n.payload,
                    "sent_at": n.sent_at,
                },
            )
            await session.commit()
